src/commands/notify.rs

```diff
@@ -0,0 +1,710 @@
+        if let Some(soa_version) = self.soa_version {
+            let soa = Record::new(
+                &self.zone,
+                Class::IN,
+                Ttl::from_secs(3600),
+                Soa::new(
+                    Name::root_vec(),
+                    Name::root_vec(),
+                    Serial(soa_version),
+                    Ttl::ZERO,
+                    Ttl::ZERO,
+                    Ttl::ZERO,
+                    Ttl::ZERO,
+                ),
+            );
+            msg.push(soa)
+                .map_err(|e| format!("could not add SOA record: {e}"))?;
+        }
+
+        let msg = msg.additional();
+
+        let tsig = self
+            .tsig
+            .as_ref()
+            .map(|tsig| {
+                Key::new(tsig.algorithm, &tsig.key, tsig.name.clone(), None, None)
+                    .map_err(|e| format!("TSIG key is invalid: {e}"))
+            })
+            .transpose()?;
+
+        let msg = msg.into_message();
+
+        writeln!(env.stdout(), "# Sending packet:");
+        writeln!(env.stdout(), "{}", msg.display_dig_style());
+
+        if self.debug {
+            writeln!(env.stdout(), "Hexdump of notify packet:");
+            writeln!(env.stdout(), "{}", base16::encode_display(&msg));
+        }
+
+        let resolver = env.stub_resolver().await;
+
+        for server in &self.servers {
+            writeln!(env.stdout(), "# sending to {}", server);
+
+            // The specified server might be an IP address. In ldns, this case is
+            // handled by `getaddrinfo`, but we have to do it ourselves.
+            // We parse it as an IP address and then send it to the one socket we
+            // can.
+            if let Ok(addr) = server.parse() {
+                let socket = SocketAddr::new(addr, self.port);
+                self.notify_host(&env, socket, msg.clone(), server, &tsig)
+                    .await;
+                continue;
+            }
+
+            let Ok(name) = Name::<Vec<u8>>::from_str(server) else {
+                writeln!(
+                    env.stderr(),
+                    "warning: invalid domain name \"{server}\", skipping."
+                );
+                continue;
+            };
+
+            let Ok(hosts) = resolver.lookup_host(&name).await else {
+                writeln!(
+                    env.stderr(),
+                    "warning: could not resolve host \"{name}\", skipping."
+                );
+                continue;
+            };
+
+            if hosts.is_empty() {
+                writeln!(
+                    env.stderr(),
+                    "skipping bad address: {name}: Name or service not known"
+                );
+                continue;
+            }
+
+            for socket in hosts.port_iter(self.port) {
+                self.notify_host(&env, socket, msg.clone(), server, &tsig)
+                    .await;
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn notify_host(
```

Comment:
Could you document what this and the other internal functions do?

Reply:
There's not a lot to document, I think. I added something, but I'm not sure what you'd like to see. What should I write?

Reply:
Even a one-line note is enough.  As I'm reading code, I like having a quick summary idea of what each high-level function does ... and I couldn't figure out one for this function, at least not at a glance.  Something simple like "Send a notify packet to a specific server"?